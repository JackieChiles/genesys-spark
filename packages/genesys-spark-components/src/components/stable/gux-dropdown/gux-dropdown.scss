@mixin gux-input-focus-border {
  outline: var(--gse-ui-formControl-input-focus-border-width)
    var(--gse-ui-formControl-input-focus-border-style)
    var(--gse-ui-formControl-input-focus-border-color);
  outline-offset: var(--gse-semantic-focusRing-offset);
}

:host {
  color: var(--gse-ui-formControl-input-populatedColor);
  box-sizing: border-box;
}

.gux-field,
.gux-target-container-expanded {
  all: unset;
  box-sizing: border-box;
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  align-content: stretch;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: var(--gse-ui-formControl-input-textfield-height);
  font-family: var(--gse-ui-formControl-input-contentText-fontFamily);
  font-size: var(--gse-ui-formControl-input-contentText-fontSize);
  line-height: var(--gse-ui-formControl-input-contentText-lineHeight);
  font-weight: var(--gse-ui-formControl-input-contentText-fontWeight);
  cursor: pointer;
  background-color: var(--gse-ui-formControl-input-backgroundColor);
}

.gux-target-container-expanded,
.gux-target-container-collapsed .gux-field {
  padding: var(--gse-ui-formControl-input-padding);
}

.gux-error {
  &.gux-target-container-collapsed .gux-field-button,
  &.gux-target-container-expanded {
    border: var(--gse-ui-formControl-input-error-border-width)
      var(--gse-ui-formControl-input-error-border-style)
      var(--gse-ui-formControl-input-error-border-color);
  }
}

.gux-disabled {
  &.gux-target-container-collapsed .gux-field-button,
  &.gux-target-container-expanded {
    border: var(--gse-ui-formControl-input-disabled-border-width)
      var(--gse-ui-formControl-input-disabled-border-style)
      var(--gse-ui-formControl-input-disabled-border-color);
  }
}

.gux-target-container-collapsed .gux-field-button:hover,
.gux-target-container-expanded:hover {
  border: var(--gse-ui-formControl-input-hover-border-width)
    var(--gse-ui-formControl-input-hover-border-style)
    var(--gse-ui-formControl-input-hover-border-color);
}

.gux-field {
  &.gux-input-field {
    height: calc(
      var(--gse-ui-formControl-input-textfield-height) -
        (var(--gse-ui-formControl-input-default-border-width) * 2)
    );
  }

  .gux-field-content {
    // COMUI-1650 consider updating default z-index values
    // Default value for popup is 2, and this tooltip must show on top
    --gux-zindex-tooltip: 3;

    display: flex;
    flex: 1 1 0;
    flex-direction: row;
    flex-wrap: nowrap;
    align-content: stretch;
    align-items: flex-start;
    justify-content: flex-start;
    min-width: 0;
    height: 60%; // keeps input centered while still visible

    .gux-filter,
    .gux-selected-option,
    .gux-placeholder {
      flex: 1 1 auto;
      align-self: auto;
      order: 0;
      padding: 0;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }

    .gux-filter {
      position: relative;
      height: 100%;

      .gux-filter-input {
        all: unset;
        width: 100%;
        color: transparent;
        caret-color: var(--gse-ui-formControl-input-populatedColor);
      }

      .gux-filter-display {
        white-space: pre;

        .gux-filter-text {
          color: var(--gse-ui-formControl-input-populatedColor);
        }

        .gux-filter-suggestion {
          opacity: 50%; // missing token ticket: COMUI-2285
        }
      }

      .gux-filter-input,
      .gux-filter-display {
        position: absolute;
      }
    }

    .gux-placeholder {
      color: var(--gse-ui-formControl-input-placeholderColor);
    }
  }

  .gux-expand-icon {
    flex: 0 0 auto;
    align-self: auto;
    order: 0;
    width: var(--gse-ui-formControl-input-inputIcon-size);
    height: var(--gse-ui-formControl-input-inputIcon-size);
    padding-left: var(--gse-ui-dropdown-gap);
    color: var(--gse-ui-formControl-input-inputIcon-iconEndColor);
  }
}

.gux-target-container-expanded {
  border: var(--gse-ui-formControl-input-active-border-width)
    var(--gse-ui-formControl-input-active-border-style)
    var(--gse-ui-formControl-input-active-border-color);
  border-radius: var(--gse-ui-formControl-input-borderRadius);

  &:focus-visible,
  &:focus-within {
    @include gux-input-focus-border;
  }

  .gux-filter-input {
    background-color: inherit;
    border: none;

    &:focus {
      border: none;
      outline: none;
    }
  }

  .gux-field-button {
    width: auto;
    height: calc(
      var(--gse-ui-formControl-input-textfield-height) -
        (var(--gse-ui-formControl-input-default-border-width) * 2)
    );
    margin: 0;
    background: inherit;
    border: none;
    outline: none;
    box-shadow: none;

    &:focus {
      outline: none;
    }
  }
}

.gux-target-container-collapsed .gux-field-button {
  border: var(--gse-ui-formControl-input-default-border-width)
    var(--gse-ui-formControl-input-default-border-style)
    var(--gse-ui-formControl-input-default-border-color);
  border-radius: var(--gse-ui-formControl-input-borderRadius);

  &:focus-visible,
  &:focus-within {
    @include gux-input-focus-border;
  }
}

.gux-listbox-container {
  box-sizing: border-box;
  margin: 0;
}

gux-popup::part(gux-popup-container) {
  width: inherit;
}

// Selected option styles

.gux-selected-icon {
  display: flex;
  align-items: center;

  gux-icon {
    width: var(--gse-ui-formControl-input-inputIcon-size);
    height: var(--gse-ui-formControl-input-inputIcon-size);
    padding-right: var(--gse-ui-dropdown-gap);
  }
}
