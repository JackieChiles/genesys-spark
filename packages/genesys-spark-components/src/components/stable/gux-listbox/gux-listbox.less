@import (reference) '../../../style/color.less';
@import (reference) '../../../style/focus.less';

:host {
  box-sizing: border-box;
  display: block;
  max-height: 20rem;
  padding: 8px 0;
  margin: 0;
  overflow-y: auto;
  color: #2e394c;
  background: @gux-grey-100;
  border: 1px solid @gux-grey-30;
  border-radius: 4px;
  outline: none;
}

:host(:focus) {
  .gux-focus-ring();
}

.gux-message-container {
  display: flex;
  flex-direction: column;
  flex-wrap: nowrap;
  align-content: stretch;
  align-items: center;
  justify-content: center;

  .gux-no-matches {
    box-sizing: border-box;
    height: 32px;
    padding: 8px 16px;
    color: #2e394c;
  }
}
