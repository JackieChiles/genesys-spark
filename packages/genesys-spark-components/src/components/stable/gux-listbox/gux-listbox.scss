@import '../../../style/color.scss';
@import '../../../style/focus.scss';
@import '../../../style/shadows.scss';

:host {
  box-sizing: border-box;
  display: block;
  max-height: 20rem;
  padding: var(--gse-ui-menu-padding);
  margin: 0;
  overflow-y: auto;
  background: var(--gse-ui-menu-backgroundColor);
  border: var(--gse-ui-menu-border-width) var(--gse-ui-menu-border-style)
    var(--gse-ui-menu-border-color);
  border-radius: var(--gse-ui-menu-borderRadius);
  box-shadow: var(--gse-ui-menu-boxShadow-x) var(--gse-ui-menu-boxShadow-y)
    var(--gse-ui-menu-boxShadow-blur) var(--gse-ui-menu-boxShadow-spread)
    var(--gse-ui-menu-boxShadow-color);
  outline: none;
}

:host(:focus) {
  outline: var(--gse-semantic-focusRing-width) solid
    var(--gse-semantic-color-focus);
  outline-offset: var(--gse-semantic-focusRing-offset);
}

.gux-message-container {
  display: flex;
  flex-direction: column;
  flex-wrap: nowrap;
  align-content: stretch;
  align-items: center;
  justify-content: center;
  text-align: center;

  .gux-no-matches {
    box-sizing: border-box;
    height: 32px;
    font-family: var(
      --gse-ui-dropdown-menu-emptyState-subheader-text-fontFamily
    );
    font-weight: var(
      --gse-ui-dropdown-menu-emptyState-subheader-text-fontWeight
    );
    line-height: var(
      --gse-ui-dropdown-menu-emptyState-subheader-text-lineHeight
    );
    font-size: var(--gse-ui-dropdown-menu-emptyState-subheader-text-fontSize);
    color: var(--gse-ui-dropdown-menu-emptyState-subheader-foregroundColor);

    .gux-no-matches-header {
      font-family: var(
        --gse-ui-dropdown-menu-emptyState-header-text-fontFamily
      );
      font-weight: var(
        --gse-ui-dropdown-menu-emptyState-header-text-fontWeight
      );
      line-height: var(
        --gse-ui-dropdown-menu-emptyState-header-text-lineHeight
      );
      font-size: var(--gse-ui-dropdown-menu-emptyState-header-text-fontSize);
      color: var(--gse-ui-dropdown-menu-emptyState-header-foregroundColor);
    }
  }
}
