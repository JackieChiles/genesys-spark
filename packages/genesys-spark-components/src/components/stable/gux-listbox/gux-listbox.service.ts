import { ListboxOptionElement } from './options/option-types';
import { logError } from '../../../utils/error/log-error';

export function getListOptions(
  list: HTMLGuxListboxElement
): ListboxOptionElement[] {
  const listChildren = Array.from(list.children);

  const allOptions = listChildren.reduce((accumulator, child) => {
    let childOptions = [];

    if (isOptionGroup(child)) {
      childOptions = Array.from(child.children).filter(item => isOption(item));
    } else if (isOption(child)) {
      childOptions = [child];
    }

    return [...accumulator, ...childOptions] as ListboxOptionElement[];
  }, []);

  return allOptions as ListboxOptionElement[];
}

export function getAvailableListOptions(
  list: HTMLGuxListboxElement
): ListboxOptionElement[] {
  return getListOptions(list).filter(option => {
    return !option.disabled && !option.filtered;
  });
}

export function isOptionGroup(item: Element): boolean {
  return item.tagName === 'GUX-OPTION-GROUP-BETA';
}

export function isOption(item: Element): boolean {
  const optionTypes = ['GUX-OPTION', 'GUX-OPTION-ICON'];
  return optionTypes.includes(item.tagName);
}

export function hasOptionGroups(list: HTMLGuxListboxElement): boolean {
  return Array.from(list.children).some(child => isOptionGroup(child));
}

function getFirstSelectedOption(
  list: HTMLGuxListboxElement
): ListboxOptionElement {
  return getListOptions(list).find(
    option => option.selected && !option.disabled && !option.filtered
  );
}

function getActiveOption(list: HTMLGuxListboxElement): ListboxOptionElement {
  return getListOptions(list).find(option => option.active);
}

function setActiveOption(
  list: HTMLGuxListboxElement,
  element: ListboxOptionElement
): void {
  if (element) {
    getListOptions(list).forEach(option => {
      const active =
        (!option.disabled || !option.filtered) && option === element;
      option.active = active;

      if (active) {
        list.setAttribute('aria-activedescendant', option.id);
      }
    });
    element.scrollIntoView({ block: 'nearest' });
  }
}

function getListOptionByLocation(
  list: HTMLGuxListboxElement,
  location: 'next' | 'previous' | 'first' | 'last'
): ListboxOptionElement {
  const availableOptionsList = getAvailableListOptions(list);
  const activeOptionIndex = hasActiveOption(list)
    ? availableOptionsList.indexOf(getActiveOption(list))
    : -1;

  switch (location) {
    case 'first':
      return availableOptionsList[0];
    case 'last':
      return availableOptionsList.slice(-1)[0];
    case 'previous':
      return hasActiveOption(list)
        ? availableOptionsList[activeOptionIndex - 1]
        : getListOptionByLocation(list, 'first');
    case 'next':
      return hasActiveOption(list)
        ? availableOptionsList[activeOptionIndex + 1]
        : getListOptionByLocation(list, 'first');
    default:
      logError(list, 'must include a valid location');
      return null;
  }
}

export function hasActiveOption(list: HTMLGuxListboxElement): boolean {
  return Boolean(getActiveOption(list));
}

export function getSearchOption(
  list: HTMLGuxListboxElement,
  searchString: string
): ListboxOptionElement {
  return getListOptions(list).find(option => {
    return (
      (!option.disabled || !option.filtered) &&
      matchOption(option, searchString)
    );
  });
}

function setSearchOptionActive(
  list: HTMLGuxListboxElement,
  searchString: string
): void {
  const option = getSearchOption(list, searchString);

  if (option) {
    setActiveOption(list, option);
  }
}

export function clearActiveOptions(list: HTMLGuxListboxElement): void {
  getListOptions(list).forEach(option => {
    option.active = false;
  });
}

export function setInitialActiveOption(list: HTMLGuxListboxElement) {
  setActiveOption(
    list,
    getFirstSelectedOption(list) || getListOptionByLocation(list, 'first')
  );
}

export function hasPreviousOption(list: HTMLGuxListboxElement): boolean {
  if (hasActiveOption(list)) {
    return Boolean(getListOptionByLocation(list, 'previous'));
  }

  return false;
}

export function hasNextOption(list: HTMLGuxListboxElement): boolean {
  if (hasActiveOption(list)) {
    return Boolean(getListOptionByLocation(list, 'next'));
  }

  return false;
}

export function setFirstOptionActive(list: HTMLGuxListboxElement): void {
  setActiveOption(list, getListOptionByLocation(list, 'first'));
}

export function setNextOptionActive(list: HTMLGuxListboxElement): void {
  setActiveOption(list, getListOptionByLocation(list, 'next'));
}

export function setPreviousOptionActive(list: HTMLGuxListboxElement): void {
  setActiveOption(list, getListOptionByLocation(list, 'previous'));
}

export function setLastOptionActive(list: HTMLGuxListboxElement): void {
  setActiveOption(list, getListOptionByLocation(list, 'last'));
}

export function actOnActiveOption(
  list: HTMLGuxListboxElement,
  handler: (value: string) => void
): void {
  if (hasActiveOption(list)) {
    handler(getActiveOption(list).value);
  }
}

export function onClickedOption(
  option: ListboxOptionElement,
  handler: (value: string) => void
): void {
  handler(option.value);
}

let timer: NodeJS.Timeout;
let searchStringState: string = '';
// While there is less than 1s between keypresses that will be considered one search operation.
// After 1s the next keypress will be considered the start of a new search.
// This is a mimic/approximation of the native select elements functionality.
const continueSearchMaxInterval = 1000;

export function goToOption(list: HTMLGuxListboxElement, letter: string): void {
  clearTimeout(timer);
  searchStringState += letter;

  setSearchOptionActive(list, searchStringState);

  timer = setTimeout(() => {
    searchStringState = '';
  }, continueSearchMaxInterval);
}

export function matchOption(
  option: ListboxOptionElement,
  matchString: string
): boolean {
  //The text content needs to be trimmed as white space can occur around the textContent if options are populated asynchronously.
  return option.textContent
    .trim()
    .toLowerCase()
    .startsWith(matchString.toLowerCase());
}
