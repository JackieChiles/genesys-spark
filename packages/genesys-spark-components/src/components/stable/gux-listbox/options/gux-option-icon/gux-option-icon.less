@import (reference) '../option-defaults.less';

gux-option-icon {
  @icon-size: 16px;
  @line-height: 20px; // measured from browser

  .gux-option-defaults();

  align-items: flex-start;

  gux-icon {
    flex-shrink: 0;
    width: @icon-size;
    height: @icon-size;
    // This & align-items: flex-start; ensures the icon is centered on the first line of text
    margin-block-start: calc((@line-height - @icon-size) / 2);
    margin-inline-end: @gux-spacing-xs;
  }
}
