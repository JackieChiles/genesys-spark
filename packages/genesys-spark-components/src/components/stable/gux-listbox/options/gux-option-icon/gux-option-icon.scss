@use '../option-defaults';

@include option-defaults.gux-option-default;

:host {
  gux-icon {
    flex-shrink: 0;
    width: var(--gse-ui-menu-option-startIcon-width);
    height: var(--gse-ui-menu-option-startIcon-height);
  }
}
