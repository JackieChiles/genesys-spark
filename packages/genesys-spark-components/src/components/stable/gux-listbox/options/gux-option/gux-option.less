@import (reference) '../option-defaults.less';

gux-option {
  .gux-option-defaults();
}
