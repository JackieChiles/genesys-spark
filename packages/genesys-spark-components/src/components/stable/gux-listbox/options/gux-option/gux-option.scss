@import '../option-defaults';

@include gux-option-default;
