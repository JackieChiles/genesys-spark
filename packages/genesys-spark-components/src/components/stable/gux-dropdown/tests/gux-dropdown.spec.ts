import { newSpecPage } from '@test/specTestUtils';

import { GuxDropdown } from '../gux-dropdown';
import { GuxListbox } from '../../gux-listbox/gux-listbox';
import { GuxOption } from '../../gux-listbox/options/gux-option/gux-option';
import { GuxOptionIcon } from '../../gux-listbox/options/gux-option-icon/gux-option-icon';

const language = 'en';

describe('gux-dropdown', () => {
  describe('#render', () => {
    it(`should render as expected with gux-option`, async () => {
      const components = [GuxDropdown, GuxListbox, GuxOption];
      const html = `
      <gux-dropdown lang="en" value="j">
        <gux-listbox aria-label="Animals">
          <gux-option value="a">Ant<span slot="subtext">Small</span></gux-option>
          <gux-option value="b">Bat</gux-option>
          <gux-option value="c">Cat</gux-option>
          <gux-option value="d">Dog</gux-option>
          <gux-option value="e">Eel</gux-option>
          <gux-option value="f">Frog</gux-option>
          <gux-option value="g">Goat</gux-option>
          <gux-option value="h">Horse<span slot="subtext">Large</span></gux-option>
          <gux-option value="i">Ibis</gux-option>
        </gux-listbox>
      </gux-dropdown>
      `;
      const page = await newSpecPage({ components, html, language });

      expect(page.rootInstance).toBeInstanceOf(GuxDropdown);
      expect(page.root).toMatchSnapshot();
    });

    it(`should render as expected with gux-option-icon`, async () => {
      const components = [GuxDropdown, GuxListbox, GuxOptionIcon];
      const html = `
      <gux-dropdown lang="en" value="j">
        <gux-listbox aria-label="Animals">
          <gux-option-icon
          icon-name="user"
          icon-color="var(--gux-blue-60)"
          value="leonardo"
          >Leonardo<span slot="subtext">TMNT</span></gux-option-icon
        >
        <gux-option-icon icon-name="user" icon-color="#8452cf" value="donatello"
          >Donatello</gux-option-icon
        >
        <gux-option-icon
          icon-name="user"
          icon-color="rgb(234, 11, 11)"
          value="raphael"
          >Raphael</gux-option-icon
        >
        <gux-option-icon
          icon-name="user"
          icon-color="#e08915"
          icon-sr-text="Screenreader information about the icon goes here, if needed"
          value="michelangelo"
          >Michelangelo</gux-option-icon
        >
        </gux-listbox>
      </gux-dropdown>
      `;
      const page = await newSpecPage({ components, html, language });

      expect(page.rootInstance).toBeInstanceOf(GuxDropdown);
      expect(page.root).toMatchSnapshot();
    });
  });
});
