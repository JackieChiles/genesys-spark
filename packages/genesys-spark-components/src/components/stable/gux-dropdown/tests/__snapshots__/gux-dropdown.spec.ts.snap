// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-dropdown #render should render as expected with gux-option 1`] = `
<gux-dropdown lang="en" value="j">
  <mock:shadow-root>
    <gux-popup-beta>
      <div class="gux-target-container-collapsed" slot="target">
        <button aria-expanded="false" aria-haspopup="listbox" class="gux-field gux-field-button" type="button">
          <div class="gux-field-content">
            <div class="gux-placeholder">
              Select...
            </div>
          </div>
          <gux-icon class="gux-expand-icon" iconname="chevron-small-down" screenreader-text="Dropdown"></gux-icon>
        </button>
      </div>
      <div class="gux-listbox-container" slot="popup">
        <slot></slot>
      </div>
    </gux-popup-beta>
  </mock:shadow-root>
  <gux-listbox aria-label="Animals" role="listbox" tabindex="0">
    <mock:shadow-root>
      <slot></slot>
    </mock:shadow-root>
    <gux-option aria-disabled="false" aria-selected="false" id="gux-option-i" role="option" value="a">
      <!---->
      <gux-truncate-beta>
        Ant
      </gux-truncate-beta>
    </gux-option>
    <gux-option aria-disabled="false" aria-selected="false" id="gux-option-i" role="option" value="b">
      <!---->
      <gux-truncate-beta>
        Bat
      </gux-truncate-beta>
    </gux-option>
    <gux-option aria-disabled="false" aria-selected="false" id="gux-option-i" role="option" value="c">
      <!---->
      <gux-truncate-beta>
        Cat
      </gux-truncate-beta>
    </gux-option>
    <gux-option aria-disabled="false" aria-selected="false" id="gux-option-i" role="option" value="d">
      <!---->
      <gux-truncate-beta>
        Dog
      </gux-truncate-beta>
    </gux-option>
    <gux-option aria-disabled="false" aria-selected="false" id="gux-option-i" role="option" value="e">
      <!---->
      <gux-truncate-beta>
        Eel
      </gux-truncate-beta>
    </gux-option>
    <gux-option aria-disabled="false" aria-selected="false" id="gux-option-i" role="option" value="f">
      <!---->
      <gux-truncate-beta>
        Frog
      </gux-truncate-beta>
    </gux-option>
    <gux-option aria-disabled="false" aria-selected="false" id="gux-option-i" role="option" value="g">
      <!---->
      <gux-truncate-beta>
        Goat
      </gux-truncate-beta>
    </gux-option>
    <gux-option aria-disabled="false" aria-selected="false" id="gux-option-i" role="option" value="h">
      <!---->
      <gux-truncate-beta>
        Horse
      </gux-truncate-beta>
    </gux-option>
    <gux-option aria-disabled="false" aria-selected="false" id="gux-option-i" role="option" value="i">
      <!---->
      <gux-truncate-beta>
        Ibis
      </gux-truncate-beta>
    </gux-option>
  </gux-listbox>
</gux-dropdown>
`;

exports[`gux-dropdown #render should render as expected with gux-option-icon 1`] = `
<gux-dropdown lang="en" value="j">
  <mock:shadow-root>
    <gux-popup-beta>
      <div class="gux-target-container-collapsed" slot="target">
        <button aria-expanded="false" aria-haspopup="listbox" class="gux-field gux-field-button" type="button">
          <div class="gux-field-content">
            <div class="gux-placeholder">
              Select...
            </div>
          </div>
          <gux-icon class="gux-expand-icon" iconname="chevron-small-down" screenreader-text="Dropdown"></gux-icon>
        </button>
      </div>
      <div class="gux-listbox-container" slot="popup">
        <slot></slot>
      </div>
    </gux-popup-beta>
  </mock:shadow-root>
  <gux-listbox aria-label="Animals" role="listbox" tabindex="0">
    <mock:shadow-root>
      <slot></slot>
    </mock:shadow-root>
    <gux-option-icon aria-disabled="false" aria-selected="false" icon-color="var(--gux-blue-60)" icon-name="user" id="gux-option-icon-i" role="option" value="leonardo">
      <!---->
      <gux-icon decorative="" icon-name="user" style="color: var(--gux-blue-60);"></gux-icon>
      <gux-truncate-beta>
        Leonardo
      </gux-truncate-beta>
    </gux-option-icon>
    <gux-option-icon aria-disabled="false" aria-selected="false" icon-color="#8452cf" icon-name="user" id="gux-option-icon-i" role="option" value="donatello">
      <!---->
      <gux-icon decorative="" icon-name="user" style="color: #8452cf;"></gux-icon>
      <gux-truncate-beta>
        Donatello
      </gux-truncate-beta>
    </gux-option-icon>
    <gux-option-icon aria-disabled="false" aria-selected="false" icon-color="rgb(234, 11, 11)" icon-name="user" id="gux-option-icon-i" role="option" value="raphael">
      <!---->
      <gux-icon decorative="" icon-name="user" style="color: rgb(234, 11, 11);"></gux-icon>
      <gux-truncate-beta>
        Raphael
      </gux-truncate-beta>
    </gux-option-icon>
    <gux-option-icon aria-disabled="false" aria-selected="false" icon-color="#e08915" icon-name="user" icon-sr-text="Screenreader information about the icon goes here, if needed" id="gux-option-icon-i" role="option" value="michelangelo">
      <!---->
      <gux-icon icon-name="user" screenreader-text="Screenreader information about the icon goes here, if needed" style="color: #e08915;"></gux-icon>
      <gux-truncate-beta>
        Michelangelo
      </gux-truncate-beta>
    </gux-option-icon>
  </gux-listbox>
</gux-dropdown>
`;
